README.md
Template Managers
=================

Our front-end and back-end for Joomla and Magento demo templates.

![Front-end](docs/img/frontend.png)

Other screenshots:

* [Administrator back-end](docs/img/backend-admin.png)
* [Account back-end](docs/img/backend-account.png)

## Vision

The software is used to manage the templates/extensions of Joomla, Magento, Wordpress, Drupal, etc.
If your company sell, provide these kind of templates, extensions, then this software is exactly for you.

The software has been used in [production](http://demo.zootemplate.com) at [ZooTemplate](http://zootemplate.com), one of the top Joomla templates/extensions provider in the world.

## Features

**Front-end**

* Responsive front-end
* View the demo in various device sizes including mobile, tablet, laptop, desktop
* Provide multiple template filters such as year, tag, responsive, etc.

**Back-end**

* Manage packages including templates and extensions
* Manage files. You can reuse files between packages
* Manage memberships taken from Amember
* Manage administrators
* Amember account can sign in and download files from packages
* Statistic the top accounts/files/packages
* Browse download history

![Account back-end](docs/img/backend-account.png)

## Platform

The software is built on the [MEAN stack](http://mean.io) which uses the following software:

* [M - MongoDB](http://mongodb.org)
* [E - Express](http://expressjs.com)
* [A - AngularJS](http://angularjs.org)
* [N - NodeJS](http://nodejs.org)

In the details, the software listed below must be installed on your server:

Software                                        | Purpose
------------------------------------------------|----------
[Amember](http://www.amember.com)               | The membership software. Amember users can use their account name/password to sign in and download the file from our back-end
[MongoDB](http://mongodb.org)                   | Database server
[NodeJS](http://nodejs.org)                     | NodeJS web server
[Nginx](http://nginx.org)                       | A web server. You can use other one such as [Apache HTTP server](http://httpd.apache.org)
[ImageMagick](http://www.imagemagick.org)       | Generating templates/extensions thumbnails
[Redis](http://redis.io)                        | Caching storage
[forever](https://github.com/nodejitsu/forever) | Watching, running the app forever without restarting the NodeJS server when updating the source code
[npm](http://npmjs.org)                         | Installing NodeJS modules

Before going to the next step, please ensure that these software are ready. Refer to their documentation for associating installation guides.

## Installing

* Download [the latest versions](https://github.com/nghuuphuoc/templatemanager/releases)
* Unzip and copy the ```src``` directory to particular directory that you want to store the app

Below is the list of steps to install the app:

### Installing NodeJS modules

From the ```src``` directory, execute the command:

```bash
$ sudo npm install
```

The command will install NodeJS modules which are listed in ```package.json```.

### Preparing the database

From the Mongo shell, run the commands (assume that the app will use ```templatemanager_dev``` as the database):

```bash
> use templatemanager_dev;
> db.user.insert({first_name: 'Administrator', last_name: '', email: 'admin@domain.com', hashed_password: '41d4736be7061d0dd826085dd5c5c773c4703e8a', salt: '1000412025288', username: 'administrator', role: 'root', locked: false });
```

At this moment, you can use the administrator account with username as ```administrator```, password as ```admin``` to manage the data in the back-end.
Of course, when the app is ready to launch, you can sign in and change the password later.

Next, index the database by commands:

```bash
> db.download.ensureIndex({ user_name: 1, downloaded_date: 1 });
> db.file.ensureIndex({ last_download: 1, num_downloads: 1 });
> db.option.ensureIndex({ key: 1 });
> db.package.ensureIndex({ created_date: 1 });
> db.package.ensureIndex({ publish: 1 });
> db.package.ensureIndex({ slug: 1 });
> db.package.ensureIndex({ type: 1 });
> db.package.ensureIndex({ year: 1 });
> db.package.ensureIndex({ tag: 1, responsive: 1, high_resolution: 1 });
> db.user.ensureIndex({ email: 1 });
> db.user.ensureIndex({ username: 1 });
```

### Creating Amember API key

As mentioned in the [Platform](#platform) section, the app connects with Amember for verifying the account and their subscriptions.
Amember allows third parties to do this via an official module named API.

The following steps show you how to install this module:

- Sign in to your Amember back-end
- Click *Setup / Configuration* on the left
- Click the *Plugins* tab on the right
- Choose *api* module from the *Enabled modules* section
- Click the *Save* button

The next steps will generate an API key:

- Click *Remote API Permissions* on the left
- Click the *New Records* button
- Type a comment in the *Comment* field
- In the *Permissions* section, ensure the following checkboxes are ticked:

Section               | Checkboxes
----------------------|-----------
Products              | **index**
Product Billing Plans | **index**, **get**
Check User Access     | **by-login-pass**, **by-login**, **by-email**

- Click the *Save* button

Amember then generates an API key which is shown in the *Api Key* field. We will use this API key in the [Setting](#setting) section.

### Setting Nginx

The app requires 2 domains to run:

* The first one handles the main website requests including the front-end and back-end
* The second one handles template thumbnails.
This domain is set by the ```thumbs.url``` option (look at the [Setting](#setting) section for more info)

Following is an example of Nginx configuration file which assumes that the main and thumbnail websites run on the
```templatemanager.dev``` and ```thumbs.templatemanager.dev``` domains, respectively:

```
server {
    listen      80;
    server_name thumbs.templatemanager.dev;
    error_log   /path/to/log/thumbs.templatemanager.dev.error.log;
    access_log  off;

    root        /path/to/thumbs/directory;
    index       index.html;
    location = /favicon.ico { access_log off; log_not_found off; }
    location = /robots.txt { access_log off; log_not_found off; }
}

upstream templatemanager {
    server 127.0.0.1:3000;
}

server {
    listen      80;
    server_name templatemanager.dev;
    error_log   /path/to/log/templatemanager.dev.error.log;
    access_log off;

    location = /favicon.ico { access_log off; log_not_found off; }
    location = /robots.txt { access_log off; log_not_found off; }

    location / {
        # This might be changed based on the size of files that the administrator can upload
        # In our production site, we allow administrators to upload templates/extensions up to 200 Mb
        client_max_body_size 200M;

        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $http_host;
        proxy_set_header X-NginX-Proxy true;

        proxy_pass http://templatemanager/;
        proxy_redirect off;
        proxy_buffering off;

        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
    }
}
```

As you see, Nginx is used as a reverse proxy which forwards all requests (coming from port 80) to NodeJS (listening on the port 3000).
If you change this port, please change the ```PORT``` variable before [running](#running) the app.

### Setting

All the app settings are placed in the ```config/config.js``` file:

```javascript
...
module.exports = {
    development: {
    },
    test: {
        root: rootPath
    },
    production: {
        root: rootPath
    }
};
```

This config file allows you to define settings for different environments such as ```development```, ```test``` and ```production```.
The settings in the ```development``` section should be used in the developing phase.
Meanwhile the production site should use the settings in the ```production``` section.

> You can indicate the environment when running the app via ```NODE_ENV``` variable.
> By default, the app will use the ```development``` settings.

Below table describes all the settings in details:

Setting                | Default               | Description
-----------------------|-----------------------|------------
root                   |                       | The root path. Please **DO NOT** change this
session.domain         | n/a                   | The cookie domain. Leave it empty if you run the app under ```localhost:<port>```
session.secret         | n/a                   | A secret string to encrypt the session data. There are a few of free online tool for generating random key, such as [RandomKeyGen](http://randomkeygen.com)
session.lifetime       | n/a                   | The session lifetime in milliseconds
db                     | n/a                   | The MongoDB connection string: ```mongodb://<database server>/<database name>```
upload.dir             | n/a                   | The path to directory storing uploaded files. __REMEMBER__ to set this directory writable
upload.maxSize         | n/a                   | Maximum size of uploaded file in kB. ```1024 * 1024 * 20``` allows user to upload files up to 20 MB in size.
thumbs.dir             | n/a                   | The directory stores the generated thumbnails of templates/extensions. __REMEMBER__ to set this directory writable
thumbs.url             | n/a                   | Prefix URL of thumbnails
thumbs.versions.square | ```['crop', 150]```   | Define the thumbnail generation method and width of thumbnail for square size. The method can be ```crop``` or ```resize```
thumbs.versions.small  | ```['resize', 240]``` | Thumbnail generation method and width for small size
thumbs.versions.medium | ```['resize', 640]``` | Thumbnail generation method and width for medium size
amember.url            | n/a                   | The URL that runs the Amember software
amember.key            | n/a                   | The Amember API key which is generated in the [Creating Amember API key](#creating-amember-api-key) section
provider.name          | n/a                   | Name of provider
provider.logo          | n/a                   | Provider logo URL
provider.url           | n/a                   | Provider URL
provider.registerUrl   | n/a                   | The register URL
redis.host             | ```localhost```       | The host of Redis server
redis.port             | 6379                  | The port of Redis server
redis.namespace        | n/a                   | Root namespace. All the Redis keys are combination of the namespace, keys, and colons (```:```)
url.frontEnd           | n/a                   | The front-end URL
url.download           | n/a                   | The download URL. In most cases, it is the same as ```url.frontEnd```. Otherwise, if the download and front-end URLs are different but have the same root domain, such as ```download.templatemanager.dev``` and ```templatemanager.dev```, please set the ```session.domain``` to root domain (```.templatemanager.dev```)

### Running

After completing all the steps above, the app is ready to launch.

Since the app is written in NodeJS, you can run it like other NodeJS app:

```bash
$ NODE_ENV=<the environment> PORT=<the listening port> node app.js
```

I also provide a shell script (```start.sh```) for running the app forever (the *forever* word means that you don't have to
restart the NodeJS server whenever you update the source code):

```
$ cd <SRC_DIRECTORY>
$ chmod 755 <SRC_DIRECTORY>/start.sh
$ nohup <SRC_DIRECTORY>/start.sh >> /var/log/template_manager.log 2>&1 &
```
(Replace ```<SRC_DIRECTORY>``` with the **full path** of ```src``` directory)

If you change the port, which is 3000 by default, remember to change the port in [Nginx](#setting-nginx) too.

## Author

This software is written by Nguyen Huu Phuoc, aka @nghuuphuoc

* [http://twitter.com/nghuuphuoc](http://twitter.com/nghuuphuoc)
* [http://github.com/nghuuphuoc](http://github.com/nghuuphuoc)

## License

```
The MIT License (MIT)

Copyright (c) 2013 - 2014 Nguyen Huu Phuoc

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```

CHANGELOG.md
Changelog
=========

## v1.2.0

Not released yet

* [#78](https://github.com/nghuuphuoc/templatemanager/issues/78): Fix "You is not allowed ..." typo
* [#80](https://github.com/nghuuphuoc/templatemanager/issues/80): Add Publish/Ubpublish function for package.
The ```package``` collection now has a new field named ```publish```:

```
> use templatemanager_dev;
> db.package.update({}, {$set: {'publish': true}}, {multi: true});
> db.package.ensureIndex({ publish: 1 });
```

* [#81](https://github.com/nghuuphuoc/templatemanager/issues/81): Upgrade Bootstrap to 3.1.1
* [#82](https://github.com/nghuuphuoc/templatemanager/issues/82): Activate tab based on URL hash
* [#83](https://github.com/nghuuphuoc/templatemanager/issues/83): Fix issue that jQuery is not found in the front-end

## v1.1.3 (2014-02-13)

* [#70](https://github.com/nghuuphuoc/templatemanager/issues/70): Upgrade Bootstrap to 3.1.0
* [#72](https://github.com/nghuuphuoc/templatemanager/issues/72): Get the user's IP address correctly
* [#73](https://github.com/nghuuphuoc/templatemanager/issues/73): Use Button component for package type selection, Responsive option
* [#74](https://github.com/nghuuphuoc/templatemanager/issues/74): Package listing page: Link to download log page
* [#75](https://github.com/nghuuphuoc/templatemanager/issues/75): Fix 'chanelog' typo
* [#76](https://github.com/nghuuphuoc/templatemanager/issues/76): URL helper: Does not append URI if there is no params
* [#77](https://github.com/nghuuphuoc/templatemanager/issues/77): Upgrade jQuery to 1.11.0 and 2.1.0

## v1.1.2 (2014-01-26)

* Use default Bootstrap theme
* Some UI improvements
* [#60](https://github.com/nghuuphuoc/templatemanager/issues/60): Show the download's browser agent
* [#62](https://github.com/nghuuphuoc/templatemanager/issues/62): Use Button component
* [#64](https://github.com/nghuuphuoc/templatemanager/issues/64): Add "Include downloads" option for each package file
* [#66](https://github.com/nghuuphuoc/templatemanager/issues/66): Upgrade to Bootstrap v3.0.3
* [#67](https://github.com/nghuuphuoc/templatemanager/issues/67): Set the custom menu for account's backend

## v1.1.1 (2014-01-17)

* [#55](https://github.com/nghuuphuoc/templatemanager/issues/55): In account back-end, show the newest packages first
* [#56](https://github.com/nghuuphuoc/templatemanager/issues/56): Show more template info in the front-end
* [#57](https://github.com/nghuuphuoc/templatemanager/issues/57): Cannot choose other template while loading a template
* [#58](https://github.com/nghuuphuoc/templatemanager/issues/58): Hide the Templates/Filter tabs when clicking the Toggle/Device icons
* [#59](https://github.com/nghuuphuoc/templatemanager/issues/59): Show the current year in footer

## v1.1.0 (2013-12-19)

* Require Redis to cache data (account memberships, tags)
* [#3](https://github.com/nghuuphuoc/templatemanager/issues/3): Auto-complete when adding tags
* [#46](https://github.com/nghuuphuoc/templatemanager/issues/46): Show the account's memberships when hovering the account name
* [#49](https://github.com/nghuuphuoc/templatemanager/issues/49): In the production env, remove debug info
* [#50](https://github.com/nghuuphuoc/templatemanager/issues/50): Custom 404, 500 error pages
* [#51](https://github.com/nghuuphuoc/templatemanager/issues/51): Nothing happen when typing the account name and clicking Enter
* [#52](https://github.com/nghuuphuoc/templatemanager/issues/52): Clear subscriptions from session data after signing out
* [#53](https://github.com/nghuuphuoc/templatemanager/issues/53): On the front-end, sort the templates by publishing year, not uploaded date

## v1.0.0 (2013-12-02)

**Front-end**

* Responsive front-end
* View the demo in various device sizes including mobile, tablet, laptop, desktop
* Provide multiple template filters such as year, tag, responsive, etc.

**Back-end**

* Manage packages including templates and extensions
* Manage files
* Manage memberships taken from Amember
* Manage administrators
* Amember account can sign in and download files from packages
* Statistic the top accounts/files/packages
* Browse download history


src/config/config.js
/**
 * Templates/Extensions manager. Powered by MEAN stack
 *
 * @link    http://github.com/nghuuphuoc/templatemanager
 * @author  http://twitter.com/nghuuphuoc
 */

var path     = require('path'),
    rootPath = path.normalize(__dirname + '/..');

module.exports = {
    development: {
        root: rootPath,
        session: {
            domain: '.templatemanager.dev',
            secret: 'XrQ2Vsw2tESughz71l1B80NwqxA7z499',
            lifetime: 3600 * 1000 * 1           // 1 hour
        },
        db: 'mongodb://localhost/templatemanager_dev',
        upload: {
            dir: '/Volumes/data/projects_workspace/templatemanager',
            maxSize: 1024 * 1024 * 1024 * 2     // 2 GB
        },
        thumbs: {
            dir: '/Volumes/data/projects_workspace/templatemanager/thumbs',
            url: 'http://thumbs.templatemanager.dev',
            versions: {
                square: ['crop', 150],
                small: ['resize', 240],
                medium: ['resize', 640]
            }
        },
        amember: {
            url: 'http://amember.dev',
            key: 'CM1ttnc4YlVEf8QhlXxS'
        },
        provider: {
            name: 'ZooTemplate',
            logo: '/asset/img/logo.png',
            url: 'http://www.zootemplate.com',
            registerUrl: 'http://www.zootemplate.com/member/signup.php'
        },
        url: {
            frontEnd: 'http://templatemanager.dev',
            download: 'http://download.templatemanager.dev'
        },
        redis: {
            host: 'localhost',
            port: 6379,
            namespace: 'templatemanager'
        }
    },
    test: {
        root: rootPath
    },
    production: {
        root: rootPath
    }
};